### PR TITLE
Added onDismiss option in createInfoNotice

### LINKED ISSUE #31571: "onDismiss" option in "createInfoNotice" is not taken into account
When using `wp.data.dispatch( 'core/notices' ).createInfoNotice` and trying to pass a function for the `onDismiss` option, the function is not triggered when the notice is dismissed.

Typing something like `wp.data.dispatch( 'core/notices' ).createInfoNotice( 'hello', { onDismiss: () => { console.log( 'dismissed!' ) } } );` is not logging anything in the console, when the created notice is dismissed.

Further information can be found [here](https://app.slack.com/client/T024MFP4J/C02QB2JS7/thread/C02QB2JS7-1620314773.351600).

### PR DESCRIPTION
**Pull request by thisissandip**
Fixes https://github.com/WordPress/gutenberg/issues/31571

## Description
Added onDismiss option in createInfoNotice method of Notices

## How has this been tested?
Create a Notice with onDismiss option
```
wp.data.dispatch( 'core/notices' ).createInfoNotice( 'hello world', { onDismiss: () => { console.log( 'hello world dismissed!' ) } } );
```

Create Multiple Notices with their onDismiss options
```
wp.data.dispatch( 'core/notices' ).createInfoNotice( 'hello 1', { onDismiss: () => { console.log( 'dismissed 1 !' ) } } );
wp.data.dispatch( 'core/notices' ).createInfoNotice( 'hello 2', { onDismiss: () => { console.log( 'dismissed 2 !' ) } } );
wp.data.dispatch( 'core/notices' ).createInfoNotice( 'hello 3', { onDismiss: () => { console.log( 'dismissed 3 !' ) } } );
```
## Screenshots

https://user-images.githubusercontent.com/69596988/120113804-d1443f80-c199-11eb-89d3-852c355ab2d0.mov

## Types of changes
Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Review by ntsekouras — COMMENTED** [review]
Thanks for working on this!

I think the place to handle this is at a single `Notice` here: https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/notice/index.js#L139

Extract the `onDismiss` prop and make sure to update the `readme`. Also a similar comment like the `snackbar` would be good: https://github.com/WordPress/gutenberg/pull/26952/files#diff-c57d338b6469546edf14bc02eb3a857fd3c24d6d6ee06b4082203ffc8ed1ef8bR52

**Comment by thisissandip** [comment]
@ntsekouras Thank you for taking a look. I made the required changes. Let me know if this looks good!

File: packages/components/src/notice/index.js [hunk 1]
```diff
@@ -84,6 +88,16 @@ function Notice( {
 		children = <RawHTML>{ children }</RawHTML>;
 	}
 
+	onDismiss = onDismiss || noop;
+
+	const dismissMe = ( event ) => {
+		if ( event && event.preventDefault ) {
```
Comment (ntsekouras) [inline packages/components/src/notice/index.js]: We can simplify with `event?.preventDefault?.();`

File: packages/components/src/notice/index.js [hunk 2]
```diff
@@ -84,6 +88,16 @@ function Notice( {
 		children = <RawHTML>{ children }</RawHTML>;
 	}
 
+	onDismiss = onDismiss || noop;
```
Comment (ntsekouras) [inline packages/components/src/notice/index.js]: Instead of doing this extra check, I think it's better to remove the default `null` value set here: https://github.com/WordPress/gutenberg/blob/trunk/packages/notices/src/store/actions.js#L63. I can't think of any reason that having `null` explicitly set is useful 🤔 

With the above change you'll need to update some tests that expect the `null`.

File: packages/components/src/notice/index.js [hunk 3]
```diff
@@ -84,6 +88,16 @@ function Notice( {
...
+
+	const dismissMe = ( event ) => {
```
Comment (ntsekouras) [inline packages/components/src/notice/index.js]: I know this is the same name with the other PR, but let's make it a bit more self explaining with something like `onDismissNotice` or something like that.

**Review by ntsekouras — COMMENTED** [review]
Thanks for following up @thisissandip ! 👍 

I've left some minor comments and after that it would be good to land!

File: packages/components/src/notice/index.js (hunk 2 above)
Comment (thisissandip) [reply packages/components/src/notice/index.js]: After removing the default null value, I ran the tests for notices as well as snackbar and it works fine. 
The null value was doing nothing since onDismiss is also handled independently in snackbar and now in notices. 😄

Comment (ntsekouras) [reply packages/components/src/notice/index.js]: Yes. You'll have to update the failing tests though that expect `onDismiss === null` to `undefined`.

Comment (thisissandip) [reply packages/components/src/notice/index.js]: I just updated the tests. Sorry for the late reply.

**Review by ntsekouras — APPROVED** [review]
Thanks for your work @thisissandip!

**Comment by thisissandip** [comment]
@ntsekouras Thank you for reviewing and merging this PR! 😄

**Comment by E-VANCE** [comment]
<s>Just a quick question:</s>

Why does using the `onDismiss`-callback only work for `createInfoNotice` but not for `createWarningNotice` or the parent `createNotice`-function that is being set up as `'info'`?

Would expect to see this being unified accross all notice-types and the wrapper function...

**UPDATE** Nevermind, the `onDismiss`-callback works for the regular `createNotice`-function just fine, must have been some issue I created for myself 😏
